### PR TITLE
Add github login function

### PR DESCRIPTION
**Pull request by 438561537**
### Description :

1. Add github login function. Store the user‘s username, avator, email into the database for future management.
2. Delete the login page.
3. Change the antd frame and route method.
4. Add syntax checks and validations of models. Refer to #68 

### Screenshot of changes :

Screenshot
check

### Please ensure Checkmark the following :

- [x] - Tested Localy
- [x] - Checked in IE11 and Firefox
- [ ] - Checked in mobile
- [x] - Backend work
- [x] - Frontend work
- [x] - Bug Fixes

### CODE REVIEW

**Comment by 438561537** [comment]
@hsluoyz  @BetaCat0  @nodece Sorry for a such big pr. Please review!

File: controllers/account.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+	githubOauthConfig.RedirectURL = RedirectURL
+	fmt.Print(RedirectURL, "\n")
+
+	// https://github.com/golang/oauth2/issues/123#issuecomment-103715338
+	// ctx := context.WithValue(oauth2.NoContext, oauth2.HTTPClient, httpClient)
+	token, err := githubOauthConfig.Exchange(context.Background(), code)
+	if err != nil {
+		res.IsAuthenticated = false
+		panic(err)
```
Comment (BetaCat0) [inline controllers/account.go:70]: Please avoid using `panic` function in coding, maybe you can log error here and interpret to user end in a friendly way.

File: controllers/account.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+		return
+	}
+	var wg sync.WaitGroup
+	var tempUserEmail []userEmailFromGithub
+	var tempUserAccount userInfoFromGithub
+	wg.Add(2)
+	go func() {
+		response, err := http.Get("https://api.github.com/user/emails?access_token=" + token.AccessToken)
+		if err != nil {
+			panic(err)
```
Comment (BetaCat0) [inline controllers/account.go:85]: Same as above

File: controllers/account.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+		}
+		defer response.Body.Close()
+		contents, err := ioutil.ReadAll(response.Body)
+		err = json.Unmarshal(contents, &tempUserEmail)
+		if err != nil {
+			res.IsAuthenticated = false
+			panic(err)
```
Comment (BetaCat0) [inline controllers/account.go:92]: Same as above

File: controllers/account.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+				res.Email = v.Email
+				break
+			}
+		}
+		wg.Done()
+	}()
+	go func() {
+		response2, err := http.Get("https://api.github.com/user?access_token=" + token.AccessToken)
+		if err != nil {
+			panic(err)
```
Comment (BetaCat0) [inline controllers/account.go:105]: ↑

File: controllers/account.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+		}
+		defer response2.Body.Close()
+		contents2, err := ioutil.ReadAll(response2.Body)
+		err = json.Unmarshal(contents2, &tempUserAccount)
+		if err != nil {
+			res.IsAuthenticated = false
+			panic(err)
```
Comment (BetaCat0) [inline controllers/account.go:112]: ↑

File: controllers/account.go [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+		res.Addition = tempUserAccount.Login
+		res.Avatar = tempUserAccount.AvatarUrl
+		if Github == "438561537" || Github == "hsluoyz" || Github == "hsluoyz" || Github == "BetaCat0" {
+			res.IsAdmin = true
+		} else {
+			res.IsAdmin = false
+		}
+		resp = Response{Status: "ok", Msg: "success", Data: res}
+	}
+	// } else {
```
Comment (BetaCat0) [inline controllers/account.go:139]: Don't leave commented codes in PR,  please think of keeping them in your local changes if necessary.

File: object/user.go [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+	IsAdmin bool   `json:"isAdmin"`
+}
+
+type UserList []*User
+
+func (list UserList) Len() int {
+	return len(list)
+}
+
+// func (list UserList) Less(i, j int) bool {
```
Comment (BetaCat0) [inline object/user.go:32]: ↑

File: object/user.go [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+func dropUserTable() error {
+	return ormManager.engine.DropTables(new(User))
+}
+
+func LinkUserAccount(field, value string) bool {
+	affected, err := ormManager.engine.Table(new(User)).Insert(map[string]interface{}{field: value})
+	if err != nil {
+		panic(err)
```
Comment (BetaCat0) [inline object/user.go:103]: Mentioned above

File: web/src/App.js [hunk 9]
```diff
@@ -1,35 +1,245 @@
...
+    super(props);
+    this.state = {
+      classes: props,
+      selectedMenuKey: 0,
+      account: null,
+    };
+  }
+
+  componentWillMount() {
+    // this.getAccount();
```
Comment (BetaCat0) [inline web/src/App.js:24]: Mentioned above

File: web/src/App.js [hunk 10]
```diff
@@ -1,35 +1,245 @@
...
-          );
-        })}
-          <Redirect to ={dashboardRoutes[0].path} from="/dashboard"/>
-          <Redirect to ="/404" />
-        </Switch>
-        </Frame>
-  ):(
-    <Redirect to ="/login" />
-  );
+            {/* <Dropdown overlay={popMenu} trigger={['click']}>
```
Comment (BetaCat0) [inline web/src/App.js:148]: Mentioned above

File: web/src/App.js [hunk 11]
```diff
@@ -1,35 +1,245 @@
...
\ No newline at end of file
+export default withRouter(App)
+
+
+
+
+
+
+
+// import React from 'react';
```
Comment (BetaCat0) [inline web/src/App.js:214]: Mentioned above

File: web/src/Setting.js [hunk 12]
```diff
@@ -1,51 +1,47 @@
...
-    ...array.slice(j + 1)
-  ];
+    return [...array.slice(0, i), array[j], ...array.slice(i + 1, j), array[i], ...array.slice(j + 1)];
+}
+
+
+export function getGithubAuthCode(method) {
+    window.event.returnValue = false
+    window.location.href = `${Conf.config.oauth_uri}?client_id=${Conf.config.client_id}&redirect_uri=${ClientUrl}/callback/${method}&scope=${Conf.GithubAuthScope}&response_type=code&state=${Conf.GithubAuthState}`;
+    // console.log(`${Conf.GithubOauthUri}?client_id=${Conf.GithubClientId}&redirect_uri=${ClientUrl}/callback/github/${method}&scope=${Conf.GithubAuthScope}&response_type=code&state=${Conf.GithubAuthState}`)
```
Comment (BetaCat0) [inline web/src/Setting.js:42]: Mentioned above

**Review by BetaCat0 — CHANGES_REQUESTED** [review]
Thanks for PR, could you please help writing some test cases to make the app stronger? :)

File: object/user.go [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+
+func (list UserList) Swap(i, j int) {
+	list[i], list[j] = list[j], list[i]
+}
+
+func GetUser(user string) *User {
+	u := User{Github: user}
+	has, err := ormManager.engine.Get(&u)
+	if err != nil {
+		panic(err)
```
Comment (BetaCat0) [inline object/user.go:38]: Avoid using `panic` in coding

File: object/user.go [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+	} else {
+		return nil
+	}
+}
+
+func IsUserAdmin(user string) bool {
+	objUser := User{Github: user}
+	has, err := ormManager.engine.Get(&objUser)
+	if err != nil {
+		panic(err)
```
Comment (BetaCat0) [inline object/user.go:52]: Ths same as above

**Review by BetaCat0 — CHANGES_REQUESTED** [review]

**Comment by BetaCat0** [comment]
I found the demo site will always notice me to login even I have done. I'm not sure if it was a bug or not. Could you help confirm?

File: web/src/pages/dashboard/ModelPage/EditModel.js [hunk 15]
```diff
@@ -27,6 +27,67 @@ class EditModel extends React.Component {
...
+
+    parseModel(text) {
+
+        if (text === null || text === undefined  || text === "") {
+            return null
+        }
+        const res = {};
+        const lines = text.match(/[^\r\n]+/g);
+        lines.forEach((line, i) => {
+            if (line.startsWith("p = ")) {
```
Comment (nodece) [inline web/src/pages/dashboard/ModelPage/EditModel.js:43]: Consider the case where there are no spaces, such as `p=`.

File: web/src/pages/dashboard/ModelPage/EditModel.js [hunk 16]
```diff
@@ -27,6 +27,67 @@ class EditModel extends React.Component {
...
+                res.p = line.slice(4);
+                res.p = this.parseLine(res.p);
+            } else if (line.startsWith("r = ")) {
```
Comment (nodece) [inline web/src/pages/dashboard/ModelPage/EditModel.js:46]: Consider the case where there are no spaces, such as `r=`

File: web/src/pages/dashboard/ModelPage/EditModel.js [hunk 17]
```diff
@@ -27,6 +27,67 @@ class EditModel extends React.Component {
...
+                res.r = line.slice(4);
+                res.r = this.parseLine(res.r);
+            } else if (line.endsWith("= _, _")) {
```
Comment (nodece) [inline web/src/pages/dashboard/ModelPage/EditModel.js:49]: Consider the case where there are no spaces, such as `=_, _`

File: web/src/pages/dashboard/ModelPage/EditModel.js [hunk 18]
```diff
@@ -27,6 +27,67 @@ class EditModel extends React.Component {
...
+                if (res.g === undefined) {
+                    res.g = [];
+                }
+                res.g.push(line.split("=")[0].trim(" "));
+            } else if (line.startsWith("e = ")) {
```
Comment (nodece) [inline web/src/pages/dashboard/ModelPage/EditModel.js:54]: Consider the case where there are no spaces, such as `e=`

File: web/src/pages/dashboard/ModelPage/EditModel.js [hunk 19]
```diff
@@ -27,6 +27,67 @@ class EditModel extends React.Component {
...
+                res.e = line.slice(4);
+            } else if (line.startsWith("m = ")) {
```
Comment (nodece) [inline web/src/pages/dashboard/ModelPage/EditModel.js:56]: Consider the case where there are no spaces, such as `m=`

File: web/src/pages/dashboard/ModelPage/AddModel.js [hunk 20]
```diff
@@ -28,30 +30,97 @@ class AddModel extends React.Component {
...
+    parseLine(s) {
+        const res = s.split(",").map(value => value.trim(" "));
+        return res;
+    }
+
+    parseModel(text) {
+        const res = {};
+        const lines = text.match(/[^\r\n]+/g);
+        lines.forEach((line, i) => {
+            if (line.startsWith("p = ")) {
```
Comment (nodece) [inline web/src/pages/dashboard/ModelPage/AddModel.js:42]: Consider the case where there are no spaces.

File: web/src/pages/dashboard/ModelPage/AddModel.js [hunk 21]
```diff
@@ -28,30 +30,97 @@ class AddModel extends React.Component {
...
+                res.r = line.slice(4);
+                res.r = this.parseLine(res.r);
+            } else if (line.endsWith("= _, _")) {
+                if (res.g === undefined) {
+                    res.g = [];
+                }
+                res.g.push(line.split("=")[0].trim(" "));
+            } else if (line.startsWith("e = ")) {
+                res.e = line.slice(4);
+            } else if (line.startsWith("m = ")) {
```
Comment (nodece) [inline web/src/pages/dashboard/ModelPage/AddModel.js:55]: Consider the case where there are no spaces.

File: web/src/pages/dashboard/ModelPage/AddModel.js [hunk 22]
```diff
@@ -28,30 +30,97 @@ class AddModel extends React.Component {
...
+                res.p = this.parseLine(res.p);
+            } else if (line.startsWith("r = ")) {
+                res.r = line.slice(4);
+                res.r = this.parseLine(res.r);
+            } else if (line.endsWith("= _, _")) {
+                if (res.g === undefined) {
+                    res.g = [];
+                }
+                res.g.push(line.split("=")[0].trim(" "));
+            } else if (line.startsWith("e = ")) {
```
Comment (nodece) [inline web/src/pages/dashboard/ModelPage/AddModel.js:53]: Consider the case where there are no spaces.

File: web/src/pages/dashboard/ModelPage/AddModel.js [hunk 23]
```diff
@@ -28,30 +30,97 @@ class AddModel extends React.Component {
...
+        const res = {};
+        const lines = text.match(/[^\r\n]+/g);
+        lines.forEach((line, i) => {
+            if (line.startsWith("p = ")) {
+                res.p = line.slice(4);
+                res.p = this.parseLine(res.p);
+            } else if (line.startsWith("r = ")) {
+                res.r = line.slice(4);
+                res.r = this.parseLine(res.r);
+            } else if (line.endsWith("= _, _")) {
```
Comment (nodece) [inline web/src/pages/dashboard/ModelPage/AddModel.js:48]: Consider the case where there are no spaces.

File: web/src/pages/dashboard/ModelPage/AddModel.js [hunk 24]
```diff
@@ -28,30 +30,97 @@ class AddModel extends React.Component {
...
+    }
+
+    parseModel(text) {
+        const res = {};
+        const lines = text.match(/[^\r\n]+/g);
+        lines.forEach((line, i) => {
+            if (line.startsWith("p = ")) {
+                res.p = line.slice(4);
+                res.p = this.parseLine(res.p);
+            } else if (line.startsWith("r = ")) {
```
Comment (nodece) [inline web/src/pages/dashboard/ModelPage/AddModel.js:45]: Consider the case where there are no spaces.

File: web/src/pages/dashboard/EnforcerPage/EditEnforcer.js [hunk 25]
```diff
@@ -14,12 +14,12 @@ class EditEnforcer extends React.Component {
...
             enforcer: null,
-            adapters: this.props.location.state[1].adapters,
-            models: this.props.location.state[1].models,
+            adapters: this.props.location.state[1],
+            models: this.props.location.state[2],
         };
     }
 
-    componentDidMount() {
+    componentWillMount() {
```
Comment (nodece) [inline web/src/pages/dashboard/EnforcerPage/EditEnforcer.js:22]: componentDidMount is correct.

File: web/src/pages/callback.js [hunk 26]
```diff
@@ -0,0 +1,50 @@
...
+  }
+
+  getAuthenticatedInfo() {
+    let redirectUrl;
+    redirectUrl = `${setting.ClientUrl}/callback/${this.state.addition}`;
+    Backend.githubLogin(this.state.code, this.state.state, redirectUrl, this.state.addition)
+      .then((res) => {
+        if (res.status === "fail") {
+          setting.showMessage("error", res?.msg);
+        }
```
Comment (nodece) [inline web/src/pages/callback.js:29]: here looks lost return.

File: web/src/App.js [hunk 27]
```diff
@@ -1,35 +1,183 @@
...
+  constructor(props) {
+    super(props);
+    this.state = {
+      classes: props,
+      selectedMenuKey: 0,
+      account: null,
+    };
+  }
+
+  componentWillMount() {
```
Comment (nodece) [inline web/src/App.js:22]: componentDidMount is correct.

File: web/src/pages/callback.js (hunk 26 above)
Comment (438561537) [reply web/src/pages/callback.js:29]: The callback page only serves as a transition to the home page, so here not needs to return.

Comment (438561537) [reply web/src/pages/callback.js:29]: @nodece Thanks for the advice. Please review.

**Review by nodece — APPROVED** [review]

**Comment by nodece** [comment]
@438561537 please fix the conflicting files.

**Comment by 438561537** [comment]
@nodece The conflicts has been removed.

**Comment by jcfergus** [comment]
Why is there no documentation on how to set up the Github OAuth application?  I have no idea what the callback URL is supposed to be.

**Comment by hsluoyz** [comment]
@438561537

**Comment by 438561537** [comment]
@jcfergus If you want to set up the Github OAuth application to run casbin-dashboard locally, you can view [this guideline](https://developer.github.com/apps/building-oauth-apps/authorizing-oauth-apps/) to set up the Github OAuth application and understand the authentication process.

**Comment by jcfergus** [comment]
Thanks for the response.  I understand how to set up OAuth - my point was that there's no documentation of this setup for casbin-dashboard to make sure that when I'm setting up my OAuth application that I use the correct callback URL and application homepage in configuring my OAuth application.  I finally found it after digging through the source (http://[hostname:port]/callback, and http://[hostname:port]/dashboard/home, respectively), but it took me rather longer to dig through to find that than it would have to either just be able to set a local password or read a quick document that told me that.   

I've decided against casbin for my particular use case, but I'll try to open a documentation pull request to at least add the above.

**Comment by 438561537** [comment]
@jcfergus Thanks for your mentioned, it‘s necessary to put more detials to explain how to configure the github OAUTH APP in the readme file.

**Comment by hsluoyz** [comment]
@438561537 please update the README about OAuth setup instructions for Casbin Dashboard, see a reference here: https://github.com/casbin/casbin-forum#installation

**Comment by 438561537** [comment]
@hsluoyz OK

**Pull request merged**